### PR TITLE
Fixed support for gcc 4.1.2.

### PR DESCRIPTION
**Pull request by JonathanHenson**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cmake/AwsCFlags.cmake [hunk 1]
```diff
@@ -16,6 +16,7 @@ include(CheckIncludeFile)
 include(CMakeParseArguments) # needed for CMake v3.4 and lower
 
 option(AWS_ENABLE_LTO "Enables LTO on libraries. Ensure this is set on all consumed targets, or linking will fail" OFF)
+option(LEGACY_COMPILER_SUPPORT "Do not use this unless you're trying to build this library on legacy compilers that you shouldn't be using in the first place" OFF)
```
Comment (bretambrose) [inline cmake/AwsCFlags.cmake]: Maybe add a more specific comment here about the actual compiler and version this fixes up: "this enables compilation for <blah blah> but versions prior to <blah blah> remain unsupported."

Comment (JonathanHenson) [reply cmake/AwsCFlags.cmake]: fair enough

File: include/aws/common/assert.h [hunk 2]
```diff
@@ -45,8 +45,10 @@ AWS_EXTERN_C_END
             bool _result = (cond);                                                                                     \
             __builtin_assume(_result);                                                                                 \
         } while (false)
-#elif defined(__GNUC__)
+#elif defined(__GNUC__) && __GNUC__ >= 4 && __GNUC_MINOR__ >= 5
```
Comment (ColdenCullen) [inline include/aws/common/assert.h]: uuuuuh so gcc 8.1 for example doesn't get this branch?

Comment (JonathanHenson) [reply include/aws/common/assert.h]: Well crap! I’ll have to do another fix.

**Pull request merged**